### PR TITLE
feat: integrate swap+send into existing tx flow

### PR DESCRIPTION
**Pull request by BZahory**
## **Description**

Swap+Send is currently decoupled from both the STX flow (we only use traditional transactions with swap+send) and the confirmations flow (we bypass this all together).

This PR reintegrates swap+send transactions with the existing STX and confirmation flows; it will also:
- Disable "edit" for swap+send transactions
- Remove the CTA button and tooltip previously used
- Update the copy on the approvals page
- Hide the failing simulation when the approval is still pending

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25734?quickstart=1)

## **Related issues**

Resolves: [METABRIDGE-904](https://consensyssoftware.atlassian.net/browse/METABRIDGE-904)

### General Flow

1. Enter the send flow on any supported swap+send network (e.g., BNB)
2. Fetch a swap+send quote from an ERC-20
3. Ensure that the "Continue" button does not submit a transaction on click or show a tooltip on hover
4. Click "Continue"
5. Ensure you can't edit the transaction
6. Ensure that the correct title copy is shown for that asset (e.g., "APPROVE USDC FOR SWAPS")
7. Submit the approval
8. Ensure that you progress to the trade transaction
9. Ensure that you can't edit the transaction
10. Ensure that you do not get a notification saying the transaction might fail
11. Ensure that the transaction can be submitted and funds are received by the recipient

### Smart Transaction Flow
1. Go to Ethereum Mainnet
2. Enter the send flow
3. Get a quote for ERC-20 -> any different asset
4. Click "Continue"
5. Confirm the approval transaction
6. Wait for the STX to complete; ensure you cannot submit the trade until it's done
7. Submit the trade and wait for the STX to complete
8. Ensure that the transaction is submitted and funds are received by the recipient

### **Before**

https://github.com/MetaMask/metamask-extension/assets/44588480/df60cb50-297d-4ad5-b60d-81a8ef2ff542

### **After**

https://github.com/MetaMask/metamask-extension/assets/44588480/29a4ae65-4dae-4863-bf65-cffd18825f4f

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js
Comment (micaelae) [inline ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:531]: Does the tx get re-simulated after the approval tx is confirmed?

Comment (BZahory) [reply ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:531]: Not that I'm aware of, but we set `isUnsimulable` to `true` if the simulation errors with a revert

**Comment by 09306677806** [comment]
0x88C478836e2f65b0322fdCB4e7f6A3e685c09e99

در تاریخ چهارشنبه ۱۰ ژوئیه ۲۰۲۴،‏ ۲۳:۴۷ sonarcloud[bot] <
***@***.***> نوشت:

**Comment by bschorchit** [comment]
For visibility, there's some [ongoing discussion](https://consensys.slack.com/archives/C068SFX90PN/p1720702160867999?thread_ts=1720616402.512989&cid=C068SFX90PN) on the reasoning for introducing the 2 confirmations instead of a single custom one.

File: ui/ducks/send/send.js
Comment (matthewwalsh0) [inline ui/ducks/send/send.js:3016]: I want to stress I think this PR is a great improvement, since we're now centralising the transaction flows through a consistent code path, confirmation, and UX.

But as a point of interest and future discussion, the send flow architecture, including the associated duck, Redux state, and draft transaction, all introduce a lot of complexity on top of the transaction code paths that can feel more a side effect of legacy code as opposed to intentional design.

If we were to implement wallet transfers for the first time today, it feels it would be far simpler, modular, and maintainable, if we were to think of a transfer as fundamentally just a standard transaction confirmation but with some additional modular React components to specify tokens and networks and amounts, and update the underlying data.

As a result, we could replace lots of the large send logic like this function, with many modular, typed and functional React components and hooks, and use them to update a pending transaction directly in the `TransactionController` state. This would mean the existing Redux state could be replaced by the React components state and the `TransactionController` unapproved transaction.

This would also align with the ongoing confirmation re-design which is based around a single confirmation template with variations, custom inputs and rows dependent on the type.

File: ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js
Comment (matthewwalsh0) [inline ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:529]:
```suggestion
    const isUnsimulatable =
```

File: ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js
Comment (matthewwalsh0) [inline ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:533]: Given the complexity and many overlapping scenarios, the `SimulationDetails` component generates its own error messages based on the result.

Why is this a special case in which the error is not shown? Is it desirable to introduce that inconsistency?

If we definitely want this, could we instead provide a new property such as `hideErrors` so we can keep the error message logic defined in one place within the component?

File: ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js
Comment (matthewwalsh0) [inline ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:1174]: Minor, but as this grows, does it warrant a `showEdit` boolean?

File: ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js
Comment (matthewwalsh0) [inline ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:1139]: To clarify, we use this in both `ui/ducks/send/send.js` and `ui/ducks/swaps/swaps.js`?

File: ui/store/actions.ts
Comment (matthewwalsh0) [inline ui/store/actions.ts:979]:
```suggestion
 * @param txParams - The transaction parameters
```

File: ui/store/actions.ts
Comment (matthewwalsh0) [inline ui/store/actions.ts:985]:
```suggestion
 * @param options.type - The type of the transaction being added.
```

File: ui/helpers/utils/transactions.util.js
Comment (matthewwalsh0) [inline ui/helpers/utils/transactions.util.js:193]: Minor, should we default this to `undefined` to make it clear it's optional?

File: ui/ducks/send/send.js
Comment (matthewwalsh0) [inline ui/ducks/send/send.js:3037]: Since we're now showing the confirmation, do we also want to run Blockaid validation as normal?

So should we update `SECURITY_PROVIDER_EXCLUDED_TRANSACTION_TYPES` in `shared/constants/security-provider.ts`?

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

**Review by cryptotavares — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
